src/int.rs

```diff
@@ -0,0 +1,261 @@
+//! This module contains an 256-bit signed integer implementation.
+
+use crate::errors::ParseI256Error;
+use std::fmt;
+use std::str::{self, FromStr};
+use web3::types::U256;
+
+/// Compute the two's complement of a U256.
+fn twos_complement(u: U256) -> U256 {
+    let (twos_complement, _) = (!u).overflowing_add(U256::one());
+    twos_complement
+}
+
+/// Little-endian 256-bit signed integer.
+#[derive(Clone, Copy, Default, Eq, Hash, PartialEq)]
+pub struct I256(U256);
+
+/// Enum to represent the sign of a 256-bit signed integer.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum Sign {
+    /// Greater than or equal to zero.
+    Positive,
+    /// Less than zero.
+    Negative,
+}
+
+impl I256 {
+    /// Creates an I256 from a sign and an absolute value. Returns the value and
+    /// a bool that is true if the conversion caused an overflow.
+    fn overflowing_from_sign_and_abs(sign: Sign, abs: U256) -> (Self, bool) {
+        let value = I256(match sign {
+            Sign::Positive => abs,
+            Sign::Negative => twos_complement(abs),
+        });
+        (value, value.sign() != sign)
+    }
+
+    /// Creates an I256 from an absolute value and a negative flag. Returns
+    /// `None` if it would overflow an `I256`.
+    fn checked_from_sign_and_abs(sign: Sign, abs: U256) -> Option<Self> {
+        let (result, overflow) = I256::overflowing_from_sign_and_abs(sign, abs);
+        if overflow {
+            None
+        } else {
+            Some(result)
+        }
+    }
+
+    /// Splits a I256 into its absolute value and negative flag.
+    fn into_sign_and_abs(self) -> (Sign, U256) {
+        let sign = self.sign();
+        let abs = match sign {
+            Sign::Positive => self.0,
+            Sign::Negative => twos_complement(self.0),
+        };
+        (sign, abs)
+    }
+
+    /// Returns the sign of self.
+    fn sign(self) -> Sign {
+        let most_significant_word = (self.0).0[3];
+        match most_significant_word & (1 << 63) {
+            0 => Sign::Positive,
+            _ => Sign::Negative,
+        }
+    }
+
+    /// Convert from a decimal string.
+    pub fn from_dec_str(value: &str) -> Result<Self, ParseI256Error> {
+        let (sign, value) = match value.as_bytes().get(0) {
+            Some(b'+') => (Sign::Positive, &value[1..]),
+            Some(b'-') => (Sign::Negative, &value[1..]),
+            _ => (Sign::Positive, value),
+        };
+
+        let abs = U256::from_dec_str(value)?;
+        let result =
+            I256::checked_from_sign_and_abs(sign, abs).ok_or(ParseI256Error::IntegerOverflow)?;
+
+        Ok(result)
+    }
+
+    /// Convert from a hexadecimal string.
+    pub fn from_hex_str(value: &str) -> Result<Self, ParseI256Error> {
+        let (sign, value) = match value.as_bytes().get(0) {
```

Comment:
Interesting. I have never seen a use case of parsing a negative hex number `-0xfff`. Maybe you want to reuse the U256:: from_hex_str and then just reuse `checked_from_sign_and_abs`?

Reply:
I think I have a note as to why we have to parse ourselves, it is because `U256::from_str` panics if the byte slice is too long.

The implementation is designed to more closely follow `i32::from_str_radix("...", 16)`.